### ISSUE TITLE
selector-attribute-quotes

### ISSUE THREAD

**Issue by evilebottnawi** [stylelint/stylelint#1137]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

possible: `"single" | "double" | "none"`

All valid:
```css
a[target="_blank"] {
    background-color: #fff;
}
```
```css
a[target='_blank'] {
    background-color: #fff;
}
```
```css
a[target=_blank] {
    background-color: #fff;
}
```

**Comment by davidtheclark** [comment]
This has been discussed a couple of times before ...

**Comment by jeddy3** [comment]
@evilebottnawi Thread of conversations start in https://github.com/stylelint/stylelint/issues/704 - have a look through and decide whether this is something you want to tackle. If not, you can close the issue.

**Comment by evilebottnawi** [comment]
> Fixes parsing of attribute selectors with equals signs in them (e.g. [data-attr="foo=bar"]) (thanks to @montmanu).
in postcss-selector-parser version 2
It is good rule, allow use `'` in `url()` functions, but not use quotes in attribute selector

**Comment by evilebottnawi** [comment]
@jeddy3 possible: "always" | "never", it does not intersect the rule `string-quotes`

**Comment by evilebottnawi** [comment]
@jeddy3 That do you think if we modify `string-quotes` for work this attribute selectors?

**Comment by davidtheclark** [comment]
So you're proposing we modify string-quotes so that it ignores attribute selectors, then add this new rule, right?

**Comment by evilebottnawi** [comment]
@davidtheclark I see two variants:
1. `string-quotes` ignore attributes selector
2. `string-quotes` add option for check attribute selector

**Comment by davidtheclark** [comment]
I think we should do the same thing we do with `url()`s (which we should document, BTW), which is just ignore them in string-quotes. If the user wants to check them, it is just as easy to say "Use this other rule" as to say "Use this option", I think.

Do we think that attribution quotes will be the only other exception — just urls and attributes? Are there other special cases for quotes lurking in the shadows?

**Comment by evilebottnawi** [comment]
@davidtheclark 
```css
a::before {
  content: ""; // <-- This?
}
```

**Comment by davidtheclark** [comment]
Why would that be special cased?

**Comment by evilebottnawi** [comment]
@davidtheclark quickly I read sorry)
We add options `checkUrl`, `checkAttributes` right?

**Comment by davidtheclark** [comment]
I do not think we should add those options. If we have separate rules to check urls and attributes, string-quotes should just ignore them rather than duplicate the behavior.

What I'm asking is: Does this stop at urls and attributes? Or are we going to end up making more exceptions and tearing string-quotes apart?

**Comment by ntwb** [comment]
> Are there other special cases for quotes lurking in the shadows?

Maybe this is one such lurker...

When using `"string-quotes": "double"` but are **not** using `font-family-name-quotes` and you have this CSS `a { font-family: 'Helvetica Neue'; }` you will get a `Expected double quotes   string-quotes` warning

**Comment by davidtheclark** [comment]
@ntwb If you *are* using `font-family-name-quotes` you still get that warning, right?

**Comment by davidtheclark** [comment]
I think the only good reason to add this rule is if there are people who want to enforce *no quotes* around attribute values. Otherwise it's just inconsistency, right?

Is that what you need, @evilebottnawi — to enforce *no quotes* around attribute values?

**Comment by evilebottnawi** [comment]
@davidtheclark yep

**Comment by jeddy3** [comment]
> I think we should do the same thing we do with url()s (which we should document, BTW), which is just ignore them in string-quotes. If the user wants to check them, it is just as easy to say "Use this other rule" as to say "Use this option", I think.

Agreed.

> Do we think that attribution quotes will be the only other exception — just urls and attributes? Are there other special cases for quotes lurking in the shadows?

I can't think of any (but see below on how we might mitigate this problem).

> Is that what you need, @evilebottnawi — to enforce no quotes around attribute values?

I think he's after single quotes for strings (including those in `url()`), but double quotes around attribute values.

> Otherwise it's just inconsistency, right?

I think that's the key issue here. Do attribute selector quotes need to match those in the HTML? If not, then it's a bit odd to add inconsistency because [they are *strings*](https://drafts.csswg.org/css-values-3/#strings).

Regarding urls and font-families. Perhaps the solution is to change the options for these rules from specifying a *type* of quote to just whether quotes are required or not e.g.

```json
{ 
  "rules": {
    "string-quotes": "double"|"single",
    "font-family-name-quotes": "where-required"|"where-recommended"|"unless-keyword",
    "function-url-quotes": "always"|"never"
  }
}
```

Again, because they [are *strings*](https://drafts.csswg.org/css-values-3/#urls). Each of those rule are standalone i.e. the font-family and url rules *just check for either type of quotes* e.g. "Expected font-family name to be quoted". A user can, optionally, use `string-quotes` to specify a type (single or double) for all strings.

Thoughts?

P.S. For `v7` obviously :)

**Comment by jeddy3** [comment]
> @davidtheclark yep

FYI, that would be `selector-attribute-quotes: "always"|"never"` using the proposal above.

**Comment by evilebottnawi** [comment]
@jeddy3 This is not a simple rule, I probably would suggest breaking it down into rules `function-url-quotes`, `selector-attrube-quotes` and etc. It may seem strange, but to maintain such rules will be much simpler to configure and will allow any detailed quotes and where to use

**Comment by ntwb** [comment]
>  If you are using `font-family-name-quotes` you still get that warning, right?

Yes, everything here works as expected, just wanted to mention a case where once one rule is removed another rule will report on the same code.

> Regarding urls and font-families. Perhaps the solution is to change the options for these rules from specifying a type of quote to just whether quotes are required or not e.g.

@jeddy3 I like your idea of the simplification of the *quotes* rules to all share the `string-quotes` single or double option, one concern I have (not that I can think of any cases off the top of my head) is that if the options are deprecated for those rules and a use case appears for where the use of single quoted strings for one rule and double quoted strings for another rule surfaces then we deprecate the new options to revert to needing the old options again.

**Comment by jeddy3** [comment]
> a use case appears for where the use of single quoted strings for one rule and double quoted strings for another rule surfaces then we deprecate the new options to revert to needing the old options again.

Yep, that is a con to this approach. It's hard to imagine ever wanting that inconsistency though.

**Comment by jeddy3** [comment]
> I probably would suggest breaking it down into rules function-url-quotes, selector-attrube-quotes and etc

In this context the original (and wide-reaching) `string-quotes` rule doesn't make sense and would need to be replaced by separate (and non-overlapping) rules. The problem with this that it could get very bitty e.g.

```css
a {
  content: "x";
}
```

What do we call the rule to target these? `content-quotes`? I guess that kind of aligned with `font-family-name-quotes`. Can anyone else think of any other occurrences of *strings* in CSS?

I think both approaches have their pros and cons, and there's no obvious right answer here.

**Comment by davidtheclark** [comment]
I really like the idea you proposed, @jeddy3, of using `string-quotes` to always enforce a single type of quotation mark, and specific rules to enforce whether urls and attribute names have quotes at all. (I do not think it's worthwhile to add complexity to allow for inconsistent quotation marks.)

**Comment by evilebottnawi** [comment]
@davidtheclark i think we should see what may have quotes and add option to ignore that users may writing plugin for special use case.

**Comment by jeddy3** [comment]
> and add option to ignore that users may writing plugin for special use case.

Lets cross that bridge when we get a real-world request.

**Comment by jeddy3** [comment]
> I really like the idea you proposed, @jeddy3,

Groovy. Lets rock that.

- Name: `selector-attribute-quotes`
- Primary options: `"always"|"never"`
- Secondary options: none
- Message: `"Expected quote" / "Unexpected quote"`
- Section: Selector
